domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,381 @@
+		archId, ok := s.architectureIDsByName[m.Arch]
+		if !ok {
+			s.logger.Warningf("skipping metadata for unknown architecture %q", m.Arch)
+			continue
+		}
+
+		// Update creation time if not defined.
+		var noTime time.Time
+		if m.CreationTime != noTime {
+			createdAt = m.CreationTime
+		}
+
+		values = append(values, inputMetadata{
+			Key:             key,
+			CreatedAt:       createdAt,
+			Source:          m.Source,
+			Stream:          m.Stream,
+			Region:          m.Region,
+			Version:         m.Version,
+			VirtType:        m.VirtType,
+			ArchitectureID:  archId,
+			RootStorageType: m.RootStorageType,
+			RootStorageSize: m.RootStorageSize,
+			Priority:        m.Priority,
+			ImageID:         m.ImageId,
+		})
+	}
+
+	// Prepare statement for checking if an image metadata already exists in DB, to handle update
+	existMetadataStmt, err := sqlair.Prepare(`
+SELECT &metadataKey.unique_key FROM cloud_image_metadata WHERE unique_key = $metadataKey.unique_key `, metadataKey{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare statement to update an existing image in DB (it only update the imageID
+	updateMetadataStmt, err := sqlair.Prepare(`
+UPDATE cloud_image_metadata
+SET image_id = $metadataImageID.image_id
+WHERE unique_key = $metadataKey.unique_key`, metadataImageID{}, metadataKey{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare statement to insert a batch of non existing image in db
+	insertMetadataStmt, err := sqlair.Prepare(`
+INSERT INTO cloud_image_metadata (*)
+VALUES ($inputMetadata.*);`, inputMetadata{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Define for each values if it should be updated or inserted
+		updated := make([]inputMetadata, 0, len(values))
+		inserted := make([]inputMetadata, 0, len(values))
+
+		for _, v := range values {
+			var ignored metadataKey
+			err := tx.Query(ctx, existMetadataStmt, metadataKey{v.Key}).Get(&ignored)
+			if dberrors.IsErrNotFound(err) {
+				inserted = append(inserted, v)
+			} else if err != nil {
+				return errors.Trace(err)
+			} else {
+				updated = append(updated, v)
+			}
+		}
+
+		// Update images metadata
+		for _, v := range updated {
+			err := tx.Query(ctx, updateMetadataStmt, metadataImageID{v.ImageID}, metadataKey{v.Key}).Run()
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
+		// Insert images metadata
+		if len(inserted) == 0 {
+			return nil
+		}
+		return errors.Trace(tx.Query(ctx, insertMetadataStmt, inserted).Run())
+	}))
+}
+
+// DeleteMetadata deletes all metadata associated with the given image ID from the database.
+func (s *State) DeleteMetadata(ctx context.Context, imageID string) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteMetadataStmt, err := sqlair.Prepare(`
+DELETE FROM cloud_image_metadata WHERE image_id = $metadataImageID.image_id`, metadataImageID{})
```

Comment:
```suggestion
DELETE 
FROM  cloud_image_metadata 
WHERE image_id = $metadataImageID.image_id`, metadataImageID{})
```

Reply:
(however for DELETE clauses, I think it's better to have

```sql
DELETE FROM table
WHERE <clauses>
```

New line after delete is overkill, i think, and clutter the reading.